src/bsoncxx/private/error_category.cpp

```diff
@@ -0,0 +1,154 @@
+                return "tried to close array, but no array was opened";
+            case error_code::k_no_document_to_close:
+                return "tried to close document, but no document was opened";
+            case error_code::k_unmatched_key_in_builder:
+                return "can't convert builder to a valid view: unmatched key";
+            case error_code::k_unset_element:
+                return "unset document::element";
+        }
+    }
+};
+
+///
+/// An error category for libbson errors in the BSON_ERROR_JSON domain.
+///
+
+class libbson_json_error final : public std::error_category {
+   public:
+    const char* name() const noexcept override {
+        return "libbson json error";
+    }
+
+    std::string message(int code) const noexcept override {
+        switch (code) {
+            case BSON_JSON_ERROR_READ_CORRUPT_JS:
+                return "attempted to parse corrupt javascript";
+            case BSON_JSON_ERROR_READ_INVALID_PARAM:
+                return "invalid parameter passed to json parser";
+            case BSON_JSON_ERROR_READ_CB_FAILURE:
+                return "reader callback failed";
+            default:
+                return "unknown libbson json error";
+        }
+    }
+};
+
+///
+/// An error category for libbson errors in the BSON_ERROR_READER domain.
+///
+class libbson_reader_error final : public std::error_category {
+   public:
+    const char* name() const noexcept override {
+        return "libbson reader error";
+    }
+
+    std::string message(int code) const noexcept override {
+        switch (code) {
+            case BSON_ERROR_READER_BADFD:
+                return "bad file descriptor";
+            default:
+                return "unknown libbson reader error";
+        }
+    }
+};
+
+///
+/// An error category for errors from an unknown libbson domain.
+///
+class libbson_unknown_error final : public std::error_category {
+   public:
+    const char* name() const noexcept override {
+        return "libbson unknown error";
+    }
+
+    std::string message(int code) const noexcept override {
+        return "unknown libbson error in unknown domain";
+    }
+};
+
+}  // namespace
+
+namespace bsoncxx {
+BSONCXX_INLINE_NAMESPACE_BEGIN
+
+const std::error_category& bsoncxx_error_category() {
+    static const bsoncxx_error_category_impl instance{};
+    return instance;
+}
+
+const std::error_category& libbson_error_category(int domain) {
+    switch (domain) {
+        case BSON_ERROR_JSON:
+            static const libbson_json_error json_category{};
+            return json_category;
+        case BSON_ERROR_READER:
+            static const libbson_reader_error reader_category{};
+            return reader_category;
+        default:
+            static const libbson_unknown_error unknown_category{};
+            return unknown_category;
+    }
+}
+
+BSONCXX_INLINE_NAMESPACE_END
+}  // namespace bsoncxx
+
+#include <bsoncxx/config/postlude.hpp>
```

Comment:
It is harmless, but you don't actually need to include the postlude in .cpp files.